Task2

Geratoptus opened the pull request.

Geratoptus committed:
Рефакторинг

1) Унифицировал скобочную нотацию
2) Сменил нотацию полей с m_ на this
3) Убрал generateMenuBar, вынеся MenuBar в отдельный класс
4) Постарался разбить энтерами логические блоки кода

Geratoptus committed:
Закрытие с подтверждением

1) Добавил метод confirmWindowClose для выхода с подтверждением
2) Распространил метод confirmWindowClose на закрытие по "X" и по кнопке в меню.
3) Локализовал yesButtonText и noButtontext

Geratoptus committed:
Небольшие изменения

1) Сделал положение и размер окон параметрами
2) Сделал окна полями
3) Заменил System.exit(0) на dispose()

Geratoptus committed:
Добавил последовательный dispose всех окон

Geratoptus committed:
Небольшие изменения

1) Вместо параметров в createGameWindow и createLogWindow сделал декоратор над addWindow с параметрами.
2) Добавил ускоритель на кнопки

Geratoptus committed:
task2

Сделал сохранение состояния окон и его подгрузку из логов.

Geratoptus committed:
Добавил запоминание свернутости окон

MaiklM commented on src/main/java/org/robots/gui/MainApplicationFrame.java:
```diff
@@ -2,155 +2,105 @@
+                MainApplicationFrame.this.confirmWindowClose();
+            }
+        });
+        stateManager.restoreState();
```
а если я этот метод вызову из какого-нибудь другого класса? он мне окна восстановит? из названия совсем не очевидно, что делает класс.

MaiklM commented on src/main/java/org/robots/state/StateManager.java:
```diff
@@ -0,0 +1,92 @@
+    public void saveState() {
+        Map<String, String> state = new HashMap<>();
+
+        state.put("mainFrameX", Integer.toString(mainFrame.getX()));
```
3 одинаковых блока - красный флаг, что настало время выделять методы

MaiklM commented on src/main/java/org/robots/state/StateManager.java:
```diff
@@ -0,0 +1,92 @@
+
+        if (state != null){
+            mainFrame.setBounds(
+                    Integer.parseInt(state.get("mainFrameX")),
```
еще 3 одинаковых блока

MaiklM commented on src/main/java/org/robots/state/StateManager.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class StateManager implements WindowState{
```
в этом классе есть проблема, если мне понадобится добавить еще 1 окно, то его придется писать нормально так кода. К тому же некоторые из окон могут иметь кастомные, особенные сохранения. Плюс вы по большому счету не требуете от окна того, чтобы оно сохранялось. У вас этот класс обязан знать про все окна в приложении, что ломает абстракцию. Попробуйте потребовать от окна, чтобы оно сохранялось

Geratoptus committed:
Доработки

1) Сделал прослойку в виде базового класса SaveableWindow
2) Убрал класс StateManager за ненадобностью
3) Добавил кнопку для сохранения
4) Сделал подтверждение сохранения при выходе из приложения

Geratoptus committed:
Поправил логику

Изменил SaveableWindow на WindowState, так как в данном случае работаем именно с этим интерфейсом.
